### PR TITLE
Better deployments

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
2-minute PR to improve deployments. Switches some tasks around to reduce downtime to a minimum.

Needs testing on staging.

### CODE REVIEW

**Comment by luisramos0** [comment]
we precompile assets after everything is working again?
what happens meanwhile if a user accesses the app while the assets are being precompiled? they will be waiting for the process to finish right?

**Comment by Matt-Yorkley** [comment]
With this change the previous assets should still be in place whilst the new assets are being compiled, so they won't be "missing" during the process.

**Comment by luisramos0** [comment]
ok, nice, we can test it in the next deploy by accessing the app while the assets are being precompiled.
if we merge this until next Tuesday I will be doing the deploy and I can try in prod.
maybe better if someone else tests it in staging.

**Comment by Matt-Yorkley** [comment]
This needs testing before merging, and the build indicates it needs a bit of tweaking.

**Comment by Matt-Yorkley** [comment]
I made some minor adjustments and now the build is green.

I deployed master to French staging and browsed / reloaded pages continuously whilst deploying and couldn't detect any changes or disruption at all. 

I think we're basically back to zero-downtime deployments with this PR. :tada:

File: roles/deploy/tasks/deploy.yml [hunk 1]
```diff
@@ -111,14 +111,14 @@
 
 # Precompile assets
 
-- name: precompile assets
+- name: precompile assets # noqa 301
```
Comment (mkllnk) [inline roles/deploy/tasks/deploy.yml]: Since this is a newly built directory, we could test for `all.js` or similar before running this.

Comment (Matt-Yorkley) [reply roles/deploy/tasks/deploy.yml]: We could use the `creates:` directive on this task to achieve that. :+1:

Comment (Matt-Yorkley) [reply roles/deploy/tasks/deploy.yml]: I've changed my mind on this. I don't think we should couple this task to a specific file existing or not in a specific path, it means if the asset pipeline changes in the future deployments will break...

**Comment by Matt-Yorkley** [comment]
> Can we compile the assets before migrating to minimise the time between database and code change?

I just tried it, but it broke the build. :disappointed: 

It looks like deployments on new servers would no longer work. Maybe we could split the deploy playbook into two different versions at some point (for new and existing servers), but it might get messy...

**Comment by mkllnk** [comment]
Not in this PR but I like the idea of splitting that a bit. It would mean that we don't have the schema loading and seeding in the deploy step which should be safer.

Did you move the seeding further down as well or just the migration?

**Comment by sauloperez** [comment]
I totally agree about what you said about migrations @mkllnk . I checked what Capistrano does and it seems we already do what's best. We basically match Capistrano's flow:

```
deploy
  deploy:starting
    [before]
      deploy:ensure_stage
      deploy:set_shared_assets
    deploy:check
  deploy:started
  deploy:updating
    git:create_release
    deploy:symlink:shared
  deploy:updated
    [before]
      deploy:bundle
    [after]
      deploy:migrate
      deploy:compile_assets
      deploy:normalize_assets
  deploy:publishing
    deploy:symlink:release
  deploy:published
  deploy:finishing
    deploy:cleanup
  deploy:finished
    deploy:log_revision
```

source: https://capistranorb.com/documentation/getting-started/flow/

**Comment by luisramos0** [comment]
tested this in FR staging. looks good, I can reload the app while assets are precompiling.

the problem is really when the old code is not compatible with the new db, isnt it?

**Comment by sauloperez** [comment]
Yep, we'll need to be aware of it and write backward-compatible migrations or the app will fail for the duration of the precompilation.

**Pull request merged**